pkg/agent/manager/cache/cache_test.go

```diff
@@ -400,62 +451,117 @@ func TestCheckSVIDCallback(t *testing.T) {
 		assert.Nil(t, svid)
+
+		return false
 	})
 	assert.Equal(t, 1, callCount)
+	assert.Empty(t, cache.staleEntries)
 
 	// called once for FOO with new SVID
 	callCount = 0
 	svids := makeX509SVIDs(foo)
-	cache.Update(&Update{
-		Bundles:             makeBundles(bundleV2),
-		RegistrationEntries: makeRegistrationEntries(foo),
-		X509SVIDs:           svids,
-	}, func(existingEntry, newEntry *common.RegistrationEntry, svid *X509SVID) {
-		callCount++
-		assert.Equal(t, "FOO", newEntry.EntryId)
-
-		// the cache should now have the entry
-		assert.Equal(t, newEntry, existingEntry)
-		if assert.NotNil(t, svid) {
-			assert.Exactly(t, svids["FOO"], svid)
-		}
+	cache.UpdateSVIDs(&UpdateSVIDs{
+		X509SVIDs: svids,
 	})
-	assert.Equal(t, 1, callCount)
 
 	// called once for FOO with existing SVID
 	callCount = 0
-	cache.Update(&Update{
+	cache.UpdateEntries(&UpdateEntries{
 		Bundles:             makeBundles(bundleV2),
 		RegistrationEntries: makeRegistrationEntries(foo),
-	}, func(existingEntry, newEntry *common.RegistrationEntry, svid *X509SVID) {
+	}, func(existingEntry, newEntry *common.RegistrationEntry, svid *X509SVID) bool {
 		callCount++
 		assert.Equal(t, "FOO", newEntry.EntryId)
 		if assert.NotNil(t, svid) {
 			assert.Exactly(t, svids["FOO"], svid)
 		}
+
+		return true
 	})
 	assert.Equal(t, 1, callCount)
+	assert.Equal(t, map[string]bool{foo.EntryId: true}, cache.staleEntries)
+}
+
+func TestGetStaleEntries(t *testing.T) {
+	cache := newTestCache()
+
+	foo := makeRegistrationEntryWithTTL("FOO", 60)
+
+	// Create svid and entry but mark it as not update
+	cache.UpdateEntries(&UpdateEntries{
+		Bundles:             makeBundles(bundleV2),
+		RegistrationEntries: makeRegistrationEntries(foo),
+	}, func(existingEntry, newEntry *common.RegistrationEntry, svid *X509SVID) bool {
+		return false
+	})
+	assert.Empty(t, cache.GetStaleEntries())
+
+	// Mark entry as stale but no SVID
+	cache.UpdateEntries(&UpdateEntries{
+		Bundles:             makeBundles(bundleV2),
+		RegistrationEntries: makeRegistrationEntries(foo),
+	}, func(existingEntry, newEntry *common.RegistrationEntry, svid *X509SVID) bool {
+		return true
+	})
+	expectedEntries := []*StaleEntry{{Entry: cache.records[foo.EntryId].entry}}
+	assert.Equal(t, expectedEntries, cache.GetStaleEntries())
+
+	// Create SVID for stale entry and update cache
+	svids := make(map[string]*X509SVID)
+	expiredAt := time.Now()
+	svids[foo.EntryId] = &X509SVID{
+		Chain: []*x509.Certificate{{NotAfter: expiredAt}},
+	}
+	cache.UpdateSVIDs(&UpdateSVIDs{
+		X509SVIDs: svids,
+	})
+	// Updating svids removes stale entries
+	assert.Empty(t, cache.GetStaleEntries())
+
+	// Update entry and mark it as stales
+	cache.UpdateEntries(&UpdateEntries{
+		Bundles:             makeBundles(bundleV2),
+		RegistrationEntries: makeRegistrationEntries(foo),
+	}, func(existingEntry, newEntry *common.RegistrationEntry, svid *X509SVID) bool {
+		return true
+	})
+
+	expectedEntries = []*StaleEntry{{
```

Comment:
Maybe add a comment:
```suggestion
        // Assert that the entry again returns as stale. This time the `ExpiresAt` field should be populated with the expiration of the SVID.
	expectedEntries = []*StaleEntry{{
```